File: LeetCode 精选 TOP 面试题/199.binary-tree-right-side-view.py
"""
日期: 2023-11-16
题目: 二叉树的右视图
链接: https://leetcode-cn.com/problems/binary-tree-right-side-view/
"""

from typing import *
from collections import *
from leetcode_utils import *
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        ans = []
        nxt = deque()
        nxt.append(root)
        cur = deque()

        while nxt:
            ans.append(nxt[-1].val)
            cur, nxt = nxt, deque()
            while cur:
                t = cur.popleft()
                nxt.append(t.left) if t.left else None
                nxt.append(t.right) if t.right else None
        
        return ans


"""
--TESTCASE BEGIN--
TestCase 0: [1,2,3,None,5,None,4]
TestCase 1: [1,None,3]
TestCase 2: []
--TESTCASE END--
""" 

print(Solution().rightSideView(create_tree([1,2,3,None,5,None,4])))

File: LeetCodeGo/199.binary-tree-right-side-view/199.binary-tree-right-side-view.go
/*
 * @Descripttion:
 * @version:
 * @Author: TDK
 * @Date: 2023-11-16 16:56:10
 * @LastEditors: TDK
 * @LastEditTime: 2023-11-16 19:24:31
 */
/* INTRODUCTION
 *日期: 2023-11-16
 *题目: 二叉树的右视图
 *链接: https://leetcode-cn.com/problems/binary-tree-right-side-view/
 */

package main

import "container/list"

//--CODE BEGIN--
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//lint:ignore U1000 Ignore unused function check
func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	type myNode struct {
		treeNode *TreeNode
		depth    int
	}
	var depth2RightVal = make(map[int]int)
	q := list.New()
	q.PushBack(&myNode{treeNode: root, depth: 0})
	maxDepth := -1

	for q.Len() != 0 {
		t := q.Front()
		q.Remove(t)
		node := t.Value.(*myNode)
		depth2RightVal[node.depth] = node.treeNode.Val
		maxDepth = max(maxDepth, node.depth)
		if node.treeNode.Left != nil {
			q.PushBack(&myNode{treeNode: node.treeNode.Left, depth: node.depth + 1})
		}
		if node.treeNode.Right != nil {
			q.PushBack(&myNode{treeNode: node.treeNode.Right, depth: node.depth + 1})
		}
	}
	ans := []int{}
	for i := 0; i <= maxDepth; i++ {
		ans = append(ans, depth2RightVal[i])
	}
	return ans
}

func max(a, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}

//--CODE END--

/*--TESTCASE BEGIN--
 *TestCase 0: {1,2,3,nil,5,nil,4}
 *TestCase 1: {1,nil,3}
 *TestCase 2: {}
 *--TESTCASE END--
 */
func main() {

}
